HPCC-19964 Std.File.Copy not preserving compression

AttilaVamos opened the pull request:
Add code to preserve compression

Fix/replace worng '&' to '&&'

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Tested manually with local and remote copy
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

jakesmith commented on dali/ft/daft.cpp:
```diff
@@ -99,6 +99,14 @@ void CDistributedFileSystem::import(IFileDescriptor * from, IDistributedFile * t
+    bool compressOutput = options->getPropBool("@compress");
+    bool preserveCompression = options->getPropBool("@preserveCompression");
+    LOG(MCdebugInfo, unknownJob, "DFS: compressInput:%d, compressOutput:%d, preserveCompression:%d ", compressInput, compressOutput, preserveCompression);
+    if (preserveCompression && compressInput)
```
What if compressOutput = true and preserveCompression = false?

I would expect output to be compressed - but this code doesn't look like it will do that, is compression turned on somewhere else or this is a bug here?

AttilaVamos replied:
It works because the target compression already set in sprayer->setTarget(to);
However it will be more cleaner solution if I merge this preserveCompression check with the existing target compression code in sprayer->setTarget(to);

jakesmith commented on dali/ft/daft.cpp:
```diff
@@ -63,7 +63,7 @@ void CDistributedFileSystem::copy(IDistributedFile * from, IDistributedFile * to
     LOG(MCdebugInfo, unknownJob, "DFS: compressInput:%d, compressOutput:%d, preserveCompression:%d ", compressInput, compressOutput, preserveCompression);
-    if (preserveCompression & compressInput)
+    if (preserveCompression && compressInput)
         sprayer->setTargetCompression(compressInput);
```
Similar to prev. comment below, what if compressOutput = true and preserveCompression = false?

AttilaVamos replied:
Same as above.

jakesmith commented:
@AttilaVamos - a couple of questions.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fc05b4f8daa2e0d5dcc26cadb4ac646ae7f6c743
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   26 sec (00:00:26) |  175 sec (00:02:55) |   61 sec (00:01:01) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1292 sec (00:21:32) |   20 sec (00:00:20) | 1609 sec (00:26:49) |

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -2637,15 +2640,11 @@ void FileSprayer::setSourceTarget(IFileDescriptor * fd, DaftReplicateMode mode)
 {
     distributedTarget.set(target);
-    compressOutput = !encryptKey.isEmpty()||target->isCompressed();
+    LOG(MCdebugInfo, unknownJob, "FileSprayer::setTarget: compressedInput:%d, compressOutput:%d, preserveCompression:%d ", compressedInput, compressOutput, preserveCompression);
```
picky: as these are bools it would be better to use boolToStr

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -2637,15 +2640,11 @@ void FileSprayer::setSourceTarget(IFileDescriptor * fd, DaftReplicateMode mode)
     distributedTarget.set(target);
-    compressOutput = !encryptKey.isEmpty()||target->isCompressed();
+    LOG(MCdebugInfo, unknownJob, "FileSprayer::setTarget: compressedInput:%d, compressOutput:%d, preserveCompression:%d ", compressedInput, compressOutput, preserveCompression);
+    compressOutput = !encryptKey.isEmpty() || target->isCompressed() || (preserveCompression && compressedInput);
```
I'm not sure this is right.
I would imply that the IDistributeFile's meta data (that's being published)compressed, is not compressed (i.e. target->isCompressed() == false).

In tests was the published compression meta info correct?

If it were correct coming into setTarget, then (preserveCompression && compressedInput) wouldn't be needed.

I think should bet set before, as the target IDistributeFile is created.

AttilaVamos replied:
Yes, it seems to use target->isCompressed() was not correct. 
However, there are two things:
1. The compress parameter doesn't populated through the chain therefore it is always false
2. Instead of 1. the compression is applied to the target file in CFileSprayEx::onCopy() and CFileSprayEx::onSprayXXXX (in ws_fsservice.cpp).
So at this point the target file compression is already set (that's why I used it, but it is a bit misleading). 
I can see two solution:
1. Restore the compress option from the target file (with isCompressed()) actually this is implemented in this push.
2. Implement the compress parameter passing through the request/call chain.

jakesmith commented:
@AttilaVamos - comment/question added.

AttilaVamos commented:
@jakesmith please re-review.

AttilaVamos commented:
@jakesmith Did you checked the generated metainfo (especially the file parts info)?

jakesmith commented:
@AttilaVamos - I did not, but compressOutput will be true with above changes.

i.e. > 1) restoring the line in FileSprayer::setTarget

Means, keeping this line:
```
compressOutput = !encryptKey.isEmpty()||target->isCompressed();
```
in FileSprayer::setTarget(IDistributedFile * target)

and target->isCompressed() will be true as explained here : https://github.com/hpcc-systems/HPCC-Platform/pull/11354#discussion_r201782661

Worth double checking everything is as you expect though.

jakesmith commented:
@AttilaVamos  I did _not_ say delete other lines in FileSprayer::setTarget()

AttilaVamos commented:
@jakesmith That means my changes in 79e149feb2846eb7d603ff05a22985cabfe90aaf was correct?

jakesmith commented:
No, from an offline discussion, the meta data may be correct, but the code has various paths to get there and isn't consistent. 

Before FileSprayer::setTarget(IDistributeFile *target) is called, 'target' should have been correctly setup so it is already compressed if required.

Such that FileSprayer::setTarget(IDistributeFile *target), should only need to do : 
    compressOutput = target->isCompressed();

That is already the case at the moment in the case of (1) and (2) below, but not (3).
1) the DFU workunit destination->isCompressed() is true. Which esp service when creating the DFU workunit will ensure if request is to create compressed spayed file.
2) If there's an encryption key. Dfurun checks to see if the dfu workunit options has an encryption key, and if it does it will turn on appropriate properties here:
```
                            if (!encryptkey.isEmpty()) {
                                fdesc->queryProperties().setPropBool("@encrypted",true);
                                fdesc->queryProperties().setPropBool("@blockCompressed",true);
                            }
```
3) Preserve compression + input compressed. The IFileDescriptor and therefore the IDistributeFile created from it are not set to compress.
Similarly to above, there should be code to check and set in this case. I think these lines required:
```
                            if (options->getPreserveCompression() && srcFile->isCompressed())
                                fdesc->queryProperties().setPropBool("@blockCompressed",true);
```
They could be an else if of the enryptKey check above.

When the fdesc is guaranteed to be correct ahead of FileSprayer::setTarget, the preserveCompression checks in filecopy.cpp can be removed.

jakesmith commented:
filecompcopy.ecl is also not quite right.
It is creating compressed files via DFU, but is then reading them back, but is forcing the declaration to \_\_COMPRESSED\_\_. : 
```
copiedcmp1 := DATASET(prefix + 'testfile_exp_copy_cmp', rec, flat, __compressed__);
copiedcmp2 := DATASET(prefix + 'testfile_cmp_copy_cmp', rec, flat, __compressed__);
copiedcmp3 := DATASET(prefix + 'testfile_cmp_copy_cmp2', rec, flat, __compressed__);
copiedcmp4 := DATASET(prefix + 'testfile_cmp_copy_cmp3', rec, flat, __compressed__);
```
That should not be necessary and may be hiding an issue, i.e. the meta data could be wrong (not compressed) - but the \_\_COMPRESSED\_\_ would be hiding that fact.

@AttilaVamos

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: fcba95658ce7a7e0019febb0a3339d40eed3106b
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   19 sec (00:00:19) |  171 sec (00:02:51) |   62 sec (00:01:02) |    6 sec (00:00:06) |   27 sec (00:00:27) | 1303 sec (00:21:43) |   18 sec (00:00:18) | 1606 sec (00:26:46) |

AttilaVamos commented:
@jakesmith please check.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -596,6 +596,7 @@ FileSprayer::FileSprayer(IPropertyTree * _options, IPropertyTree * _progress, IR
     throttleNicSpeed = 0;
     compressedInput = false;
     compressOutput = options->getPropBool(ANcompress);
+    preserveCompression = options->getPropBool("@preserveCompression");
```
unused - should be deleted.

jakesmith commented on dali/ft/filecopy.cpp:
```diff
@@ -2637,15 +2640,16 @@ void FileSprayer::setSourceTarget(IFileDescriptor * fd, DaftReplicateMode mode)
+    LOG(MCdebugInfo, unknownJob, "FileSprayer::setTarget: compressedInput:%s, compressOutput:%s, preserveCompression:%s ",
+                                boolToStr(compressedInput),
+                                boolToStr(compressOutput),
+                                boolToStr(preserveCompression));
```
preserveCompression isn't used anymore - should be removed.

jakesmith commented:
@AttilaVamos - preserveCompression is no longer used in filecopy.cpp and can be removed.

jakesmith commented:
@AttilaVamos - looks good.

Did manual tests of fileservices and eclwatch copies with/without compression and preserve all check out correctly?

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 0151be19aab4b15d21b1c56bcc0ed15e50e11151
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   15 sec (00:00:15) |  176 sec (00:02:56) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1251 sec (00:20:51) |   18 sec (00:00:18) | 1560 sec (00:26:00) |

jakesmith commented:
@AttilaVamos - ok I'll leave with you for time being to find out cause.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 19bd02a6da345df60e36ef9cdd82d21bf0ecee11
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |789 |789 |0 |
|test (thor) |720 |720 |0 |
|test (roxie) |860 |860 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  170 sec (00:02:50) |   61 sec (00:01:01) |    9 sec (00:00:09) |   30 sec (00:00:30) | 1295 sec (00:21:35) |   18 sec (00:00:18) | 1593 sec (00:26:33) |

AttilaVamos commented:
@jakesmith Remote copy issues are fixed, please check.

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1460,9 +1473,16 @@ class CDFUengine: public CInterface, implements IDFUengine
+                        {
+                            IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : foreignfdesc.get());
+                            if (options->getPreserveCompression() && srcDesc && srcDesc->isCompressed())
+                                dstFile->queryAttributes().setPropBool("@blockCompressed",true);
```
@AttilaVamos - This will get the right result, but it should be setup before it gets here.
i.e. dstFile and the IFileDescriptor is was based, as with other cases should be preset to correct compression before it gets here.

I'm assuming this was added because it doesn't get set on lines 1326-27..?
If so, why didn't it those lines?

AttilaVamos replied:
It hit those lines and crashed there because srcFile is a distributed file (for copy) and it wasn't initialized for remote copy (that's why I add pointer checks '&& srcFile' there),

jakesmith replied:
@AttilaVamos - the check should still be before and set the target fdesc (on line 1327) as with other cases.
So it should be there that it checks foreignfdesc or auxfdesc if no srcFile is available.

jakesmith replied:
this is what I think code should look like (untested):

replace lines 1326-1327
with:
```
    else if (options->getPreserveCompression())
    {
        bool dstCompressed = false;
        if (srcFile)
            dstCompressed = srcFile->isCompressed();
        else
        {
            IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : foreignfdesc.get());
            dstCompressed = srcDesc && srcDesc->isCompressed();
        }
        if (dstCompressed)
            fdesc->queryProperties().setPropBool("@blockCompressed", dstCompressed);
    }
```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 8fe200251caad7c753b40d903bbdc0147290d9b5
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |790 |790 |0 |
|test (thor) |721 |721 |0 |
|test (roxie) |861 |861 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  173 sec (00:02:53) |   61 sec (00:01:01) |    7 sec (00:00:07) |   27 sec (00:00:27) | 1306 sec (00:21:46) |   18 sec (00:00:18) | 1614 sec (00:26:54) |

jakesmith commented on dali/dfu/dfurun.cpp:
```diff
@@ -1358,11 +1373,17 @@ class CDFUengine: public CInterface, implements IDFUengine
                                 dstFile.setown(fdir.createNew(fdesc));
+                                IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : foreignfdesc.get());
+                                if (options->getPreserveCompression() && srcDesc && srcDesc->isCompressed())
+                                    dstFile->queryAttributes().setPropBool("@blockCompressed",true);
```
>> old comment :
>>@AttilaVamos - the check should still be before and set the target fdesc (on line 1327) as with other cases.
>> So it should be there that it checks foreignfdesc or auxfdesc if no srcFile is available.

jakesmith 21 minutes ago  Member
@AttilaVamos - your latest commit is still doing it late. See last comment above, re. the change should happen at around line 1327, i.e. you should be changing fdesc (the IFileDescriptor), not later after the desc has been used to create the IDistributedFile

jakesmith replied:
this is what I think code should look like (untested):

replace lines 1326-1327
with:
```
    else if (options->getPreserveCompression())
    {
        bool dstCompressed = false;
        if (srcFile)
            dstCompressed = srcFile->isCompressed();
        else
        {
            IFileDescriptor * srcDesc = (auxfdesc.get() ? auxfdesc.get() : foreignfdesc.get());
            dstCompressed = srcDesc && srcDesc->isCompressed();
        }
        if (dstCompressed)
            fdesc->queryProperties().setPropBool("@blockCompressed", dstCompressed);
    }
```

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 7fd4f9699f55908ce02ef25cedc55784a7d6684a
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |790 |790 |0 |
|test (thor) |721 |721 |0 |
|test (roxie) |861 |861 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   10 sec (00:00:10) |  173 sec (00:02:53) |   61 sec (00:01:01) |    7 sec (00:00:07) |   28 sec (00:00:28) | 1310 sec (00:21:50) |   20 sec (00:00:20) | 1609 sec (00:26:49) |

AttilaVamos commented:
@jakesmith please check

jakesmith commented:
@AttilaVamos - looks good, please squash, and rebase to latest master.

AttilaVamos committed:
HPCC-19964 Std.File.Copy not preserving compression

Add code to preserve compression

Remove unused definitions and code

Update filecompcopy.ecl

Tested maually with local and remote copy

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: eac7723c3ba36c4d40144b6596a5337bfaa987b9
Build: success
Install hpccsystems-platform-community_7.0.0-trunk0.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |790 |790 |0 |
|test (thor) |721 |721 |0 |
|test (roxie) |861 |861 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   22 sec (00:00:22) |  168 sec (00:02:48) |   62 sec (00:01:02) |    8 sec (00:00:08) |   29 sec (00:00:29) | 1290 sec (00:21:30) |   19 sec (00:00:19) | 1598 sec (00:26:38) |

AttilaVamos commented:
@jakesmith done,

jakesmith commented:
@AttilaVamos looks good

@richardkchapman - ready to merge.

richardkchapman commented on dali/ft/filecopy.cpp:
```diff
@@ -2637,15 +2640,17 @@ void FileSprayer::setSourceTarget(IFileDescriptor * fd, DaftReplicateMode mode)
 void FileSprayer::setTarget(IDistributedFile * target)
 {
     distributedTarget.set(target);
-    compressOutput = !encryptKey.isEmpty()||target->isCompressed();
```
Was the removal of the encryptKey test deliberate?

richardkchapman commented:
@AttilaVamos One question

AttilaVamos commented:
@richardkchapman yes, it was. The '!encryptkey.isEmpty()' condition already processed earlier, in dfurun.cpp:1321.
